javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java

```diff
@@ -489,8 +489,16 @@ public void onComplete(
     @Override
     public Iterable<Iterable<Class<?>>> resolve(Instrumentation instrumentation) {
       // filter out our agent classes and injected helper classes
-      return Iterables.transform(
-          delegate.resolve(instrumentation), i -> Iterables.filter(i, c -> !isIgnored(c)));
+      return () ->
```

Comment:
Hope you do understand this and tested it :) Looks like black magic to me

Reply:
The Stream (single use) vs Iterable (multiuse) dilemma is always annoying. The safe way to convert a Stream into an Iterable is creating a Lambda that initializes the Stream and returns the iterator, which is what an Iterable is.

The alternative is to collect(toList) each time which I'm ok with but I don't know this part well enough to know if the extra allocations would be problematic.

Reply:
I think naming the Iterable <-> Stream transformations makes it (a bit) easier to read:

```
    public Iterable<Iterable<Class<?>>> resolve(Instrumentation instrumentation) {
      // filter out our agent classes and injected helper classes
      return iterableOf(
          streamOf(delegate.resolve(instrumentation))
              .map(classes -> iterableOf(streamOf(classes).filter(c -> !isIgnored(c)))));
    }

    private static <E> Stream<E> streamOf(Iterable<E> iterable) {
      return StreamSupport.stream(iterable.spliterator(), false);
    }

    private static <E> Iterable<E> iterableOf(Stream<E> stream) {
      return stream::iterator;
    }
```

Reply:
Unfortunately `stream:: iterator` isn't a real Iterable since it can't be replayed, so `toIterable` would need to accept `Supplier<Stream>`. Not sure if that ends up better.

Reply:
Tried, how is it?

Reply:
> Unfortunately `stream:: iterator` isn't a real Iterable since it can't be replayed

😭

ok, how about

```
    public Iterable<Iterable<Class<?>>> resolve(Instrumentation instrumentation) {
      // filter out our agent classes and injected helper classes
      return () -> streamOf(delegate.resolve(instrumentation)).map(this::filterClasses).iterator();
    }

    private Iterable<Class<?>> filterClasses(Iterable<Class<?>> classes) {
      return () -> streamOf(classes).filter(c -> !isIgnored(c)).iterator();
    }

    private static <T> Stream<T> streamOf(Iterable<T> iterable) {
      return StreamSupport.stream(iterable.spliterator(), false);
    }
```

Reply:
Sure seems good